sound/soc/sof/intel/hda-loader.c

```diff
@@ -472,6 +472,51 @@ int hda_dsp_post_fw_run(struct snd_sof_dev *sdev)
 	return hda_dsp_ctrl_clock_power_gating(sdev, true);
 }
 
+/*
+ * post fw run operations for ICL,
+ * Core 3 will be powered up and in stall when HPRO is enabled
+ */
+int hda_dsp_post_fw_run_icl(struct snd_sof_dev *sdev)
+{
+	struct sof_intel_hda_dev *hda = sdev->pdata->hw_pdata;
+	int ret;
+
+	if (sdev->first_boot) {
+		ret = hda_sdw_startup(sdev);
+		if (ret < 0) {
+			dev_err(sdev->dev,
+				"error: could not startup SoundWire links\n");
+			return ret;
+		}
+	}
+
+	hda_sdw_int_enable(sdev, true);
+
+	/*
+	 * The recommended HW programming sequence for ICL is to
+	 * power up core 3 and keep it in stall if HPRO is enabled
+	 */
+	if (!hda->clk_config_lpro) {
+		int icl_core_mask = 1 << (hda->desc->cores_num - 1);
+
+		dev_dbg(sdev->dev, "dsp core power up: icl_core_mask %d\n", icl_core_mask);
+
+		ret = hda_dsp_core_power_up(sdev, icl_core_mask);
+		if (ret < 0) {
+			dev_err(sdev->dev, "error: dsp core power up failed, core_mask %d\n", icl_core_mask);
+			return ret;
+		}
+
+		/* update enabled_cores_mask to keep track of enabled cores */
+		sdev->enabled_cores_mask |= icl_core_mask;
+
+		snd_sof_dsp_stall(sdev, icl_core_mask);
+	}
+
+	/* re-enable clock gating and power gating */
+	return hda_dsp_ctrl_clock_power_gating(sdev, true);
+}
```

Comment:
can you clarify why you need this routine when we don't seem to need it for TGL? Is this intentional?

Reply:
@plbossart this is intentional. On ICL core 3 is managed by the host whereas on TGL it is handled by the firmware

Reply:
A comment wouldn't hurt then.

Reply:
I updated with more comments